### ISSUE TITLE
changing tiddlers in the story when changing the language

### ISSUE THREAD

**Issue by tobibeer** [Jermolene/TiddlyWiki5#2597]
As an author, providing users a means to switch the language of a TiddlyWiki, I want to have tiddlers displayed in the story replaced with their variant in the language switched to.

## Example

Imagine these tiddlers:

```
title: Willcommen

Hallo. Schön, dass Du hier bist!
lang: de-DE
```

```
title: Welcome

Hi. Nice to have you around!
lang: en-GB
```

And **default tiddlers** in the **ControlPanel** set to:

```
[{$:/language}removeprefix[$:/languages]addprefix[$:/_lang/start]get[text]]
```

...which reads the home page from whatever is stored in either of:

* **$:/_lang/start/de-DE**
* **$:/_lang/start/en-GB**

...depending on the currently selected language.

Even for the default tiddlers, I don't want a two-step process, where first the language is changed and then one needs to take an additional click to open the home page, which one might not even want.

I would simply want all tiddlers replaced with their equivalent in the other language, escept for those where none can be found.

**Comment by Jermolene** [comment]
Hi @tobibeer interesting.

If I understand you correctly, the problem you have with the approach that you outline is that if the user changes the language dynamically the currently displayed tiddlers wouldn't change?

Another approach to the multilingual content problem would be a custom `$:/core/ui/ViewTemplate/body`. Rather than multiple tiddlers with the same title, it might use the convention of suffixes like `[de-DE]` to indicate the language of a tiddler, with the lack of a suffix perhaps indicating a default such as `[en-GB]`. The behaviour might be something like:

* If there is a language suffix, then display the tiddler as is
* If there is no language suffix, then add the current language suffix; if the resulting tiddler exists, display it, otherwise display the default content (ie the current tiddler)

I'm sure there's a lot of different ways of tackling the problem :smile:

**Comment by tobibeer** [comment]
I am quite sure, I need distinct tiddlers in each language and so I need a way to replace the ones in the story. Unfortunately, it's not just a matter of ViewTemplate and what to display, One reason would be search, another tagging.

Perhaps there could be some language-switcher-actions triggered when the language is changed?

Then I could have list widgets inspect the currently set language, walk through the story and replace any that have an equivalent in that new language with exactly that via some tm-navigate action.

**Comment by Jermolene** [comment]
> I am quite sure, I need distinct tiddlers in each language and so I need a way to replace the ones in the story. Unfortunately, it's not just a matter of ViewTemplate and what to display, One reason would be search, another tagging, which need adapting as well.

OK, so it would be a custom override of `$:/core/ui/ViewTemplate`, not `$:/core/ui/ViewTemplate/body`, but I think the technique still works better than messing around with the content of the story river.

**Comment by tobibeer** [comment]
Sry, edited my last post. Perhaps read again regarding an **actions** attribute to the linkcatcher.

**Comment by Jermolene** [comment]
The trouble with the language switcher actions attribute is that it would only apply when the language is switched via the standard switcher, and not by, say, a button that changes `$:/language` directly.

But you don't need the language switcher actions if you adopt the approach I'm suggesting.

It's good to experiment on stuff like multilingual content, but it's not a great strategy to start an experiment by requiring deep changes to the core. Inevitably you'll get bogged down in discussions about the generality of the change, and meanwhile this cool area goes unexplored :smile:

**Comment by tobibeer** [comment]
Cool area, but with a workaroundish implementation.

Again, my problem are the tiddlers in the story that should not be there anymore when the language is switched... no matter what the listener.

**Comment by tobibeer** [comment]
Otherwise I would have to rewrite the entire EditTemplate and ViewTemplate to cater for multi-language.

**Comment by tobibeer** [comment]
For me, the easiest bit is to be able to control what happens if someone uses the language switcher to switch languages. That anyone could do that by editing `$:/language`, is quite a hypothetical... and in that case, I am perfectly fine, if the entire presentation doesn't entirely switch accordingly, but only the core bits.

**Comment by Jermolene** [comment]
> Otherwise I would have to rewrite the entire EditTemplate and ViewTemplate bits to cater for multi-language.

That's exactly what I'm suggesting that you should do!

> Again, my problem are the tiddlers in the story that should not be there anymore when the language is switched... no matter what the listener.

The approach I've outlined address that problem, doesn't it?

**Comment by Jermolene** [comment]
> For me, the easiest bit is to be able to control what happens if someone uses the language switcher to switch languages. That anyone could do that by editing $:/language, is quite a hypothetical... 

That's your opinion, but it's not robust enough logic for the core. I use the technique of directly writing to `$:/language` etc all the time in my TiddlyWiki projects.

Once again, please don't kick off speculative exploratory investigations by requesting core changes. It's slowing you down, and it slows me down. Why not just get on with implementing something and report back?

**Comment by tobibeer** [comment]
Your approahc works, but the amount of components to touch and enhance is quite substantial.

**Comment by tobibeer** [comment]
Will explore some.

**Comment by Jermolene** [comment]
> Your approach works, but the amount of components to touch and enhance is quite substantial.

Not at all. It requires an override of `$:/core/ui/ViewTemplate` but that is **literally** all of the core that needs changing, or touching in any way.

**Comment by tobibeer** [comment]
@Jermolene, I'm afraid with your changes navigation fails for TiddlyLinks, as the story actually doesn't hold the tiddler originally being asked to open anymore.

**Comment by Jermolene** [comment]
> navigation fails for TiddlyLinks, as the story actually doesn't hold the tiddler originally being asked to open anymore.

The story would hold the tiddlers that were navigated to; it's the display of those tiddlers that would be different. Things would get messed up if a user was allowed to navigate directly to a translated tiddler, but otherwise I think things would just work.

**Comment by tobibeer** [comment]
What I did was determine the tiddler variable for the ViewTemplate to then handle. I guess that's not what you had in mind then.

**Comment by linonetwo** [comment]
@tobibeer Seems you mention language problems many times, long before I started using Tiddlywiki. Now this is closable, hope you still reading github notificaions!
